### PR TITLE
Decouple session from connection

### PR DESCRIPTION
**Pull request by lutovich**

### CODE REVIEW

File: driver/src/main/java/org/neo4j/driver/internal/InternalDriver.java [hunk 1]
```diff
@@ -61,42 +64,57 @@ public final Session session()
...
+        Session session = sessionFactory.newInstance( mode );
+        if ( closed.get() )
         {
             // the driver is already closed and we either 1. obtain this session from the old session pool
             // or 2. we obtain this session from a new session pool
             // For 1. this closeResources will take no effect as everything is already closed.
             // For 2. this closeResources will close the new connection pool just created to ensure no resource leak.
             closeResources();
-            throw driverCloseException();
+            throw new DriverClosedException();
```
Comment (zhenlineo) [inline driver/src/main/java/org/neo4j/driver/internal/InternalDriver.java]: Maybe not a new exception type in driver for now

File: driver/src/main/java/org/neo4j/driver/internal/InternalStatementResult.java [hunk 2]
```diff
@@ -319,6 +321,7 @@ private boolean tryFetchNext()
         {
             if ( done )
             {
+                connectionHandler.resultBuffered();
```
Comment (zhenlineo) [inline driver/src/main/java/org/neo4j/driver/internal/InternalStatementResult.java]: if you do `connectionHandler.resultBuffered()` at the place where `done` is set to be true, then you probably could also address failure/ignore cases too.

File: driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java [hunk 3]
```diff
@@ -132,21 +118,23 @@ public synchronized void reset()
...
         if ( currentTransaction != null )
         {
             currentTransaction.markToClose();
             lastBookmark = currentTransaction.bookmark();
             currentTransaction = null;
         }
-        connection.resetAsync();
+        if ( currentConnection != null )
+        {
+            currentConnection.resetAsync();
```
Comment (zhenlineo) [inline driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java:124]: I would suggest to remove reset. Make this method deprecated and then give an empty impl inside.

File: driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java [hunk 4]
```diff
@@ -241,25 +203,42 @@ public TypeSystem typeSystem()
...
+        closeCurrentConnection( true );
+    }
+
+    @Override
+    public synchronized void transactionClosed( ExplicitTransaction tx )
+    {
+        if ( currentTransaction != null && currentTransaction == tx )
+        {
+            lastBookmark = currentTransaction.bookmark();
+            currentTransaction = null;
```
Comment (zhenlineo) [inline driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java]: connection also need to be closed

File: driver/src/main/java/org/neo4j/driver/internal/net/pooling/PooledSocketConnection.java [hunk 5]
```diff
@@ -58,7 +59,7 @@
     private final Consumer<PooledConnection> release;
 
     private boolean unrecoverableErrorsOccurred = false;
-    private Runnable onError = null;
+    private ConnectionHandler connectionHandler;
```
Comment (zhenlineo) [inline driver/src/main/java/org/neo4j/driver/internal/net/pooling/PooledSocketConnection.java]: suggesting rename something else such as `ConnectionTerminationHandler`?

File: driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java (hunk 3 above)
Comment (lutovich) [reply driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java:124]: Can we maybe remove it in a separate PR? This one is quite big already :)

File: driver/src/test/java/org/neo4j/driver/internal/InternalStatementResultTest.java [hunk 6]
```diff
@@ -386,23 +386,89 @@ public void shouldNotPeekIntoTheFutureWhenResultIsEmpty()
...
+        verify( resourcesHandler ).onResultConsumed();
+    }
+
+    @Test
+    public void shouldNotifyResourcesHandlerWhenSummary()
+    {
+        SessionResourcesHandler resourcesHandler = mock( SessionResourcesHandler.class );
+        StatementResult result = createResult( 10, resourcesHandler );
+
+        assertNotNull( result.summary() );
```
Comment (zhenlineo) [inline driver/src/test/java/org/neo4j/driver/internal/InternalStatementResultTest.java:433]: maybe summary several times to ensure only called once?
Similar to other methods and also some combination of all the methods? such as result.single(), result.summary(), then result.consume()?

Comment (lutovich) [reply driver/src/test/java/org/neo4j/driver/internal/InternalStatementResultTest.java:433]: Added a separate test for this here: https://github.com/neo4j/neo4j-java-driver/pull/324/commits/0ddac3d5036b7017f80f6592ba4815b0b00bcd87

**Pull request merged**